parsers/cxx/cxx_parser.c

```diff
@@ -0,0 +1,1047 @@
+	if(cxxTokenTypeIsOneOf(g_cxx.pToken,CXXTokenTypeEOF | CXXTokenTypeSemicolon))
+	{
+		CXX_DEBUG_LEAVE_TEXT("Found EOF/semicolon while parsing if/for/while/switch");
+		return TRUE;
+	}
+	
+	if(cxxTokenTypeIs(g_cxx.pToken,CXXTokenTypeParenthesisChain))
+	{
+		// Extract variables from the parenthesis chain
+		// We handle only simple cases.
+		CXXTokenChain * pChain = g_cxx.pToken->pChain;
+
+		CXX_DEBUG_ASSERT(pChain->iCount >= 2,"The parenthesis chain must have initial and final parenthesis");
+
+		// Kill the initial parenthesis
+		cxxTokenChainDestroyFirst(pChain);
+		// Fake the final semicolon
+		CXXToken * t = cxxTokenChainLast(pChain);
+		t->eType = CXXTokenTypeSemicolon;
+		vStringClear(t->pszWord);
+		vStringPut(t->pszWord,';');
+
+		// and extract variable declarations if possible
+		cxxParserExtractVariableDeclarations(pChain);
+
+		CXX_DEBUG_LEAVE_TEXT("Found if/for/while/switch parenthesis chain");
+		return TRUE;
+	}
+
+	// must be opening bracket: parse it here.
+
+	boolean bRet = cxxParserParseBlock(TRUE);
+
+	CXX_DEBUG_LEAVE();
+
+	return bRet;
+}
+
+rescanReason cxxParserMain(const unsigned int passCount)
+{
+	if(g_bFirstRun)
+	{
+		cxxTokenAPIInit();
+
+		g_cxx.pTokenChain = cxxTokenChainCreate();
+
+		cxxScopeInit();
+		
+		g_bFirstRun = FALSE;
+	} else {
+		// Only clean state
+		cxxScopeClear();
+		cxxTokenAPINewFile();
+		cxxParserNewStatement();
+	}
+
+	kindOption * kind_for_define = cxxTagGetKindOptions() + CXXTagKindMACRO;
+	kindOption * kind_for_header = cxxTagGetKindOptions() + CXXTagKindINCLUDE;
+	int role_for_macro_undef = CR_MACRO_UNDEF;
+	int role_for_header_system = CR_HEADER_SYSTEM;
+	int role_for_header_local = CR_HEADER_LOCAL;
+
+	Assert(passCount < 3);
+
+	cppInit(
+			(boolean) (passCount > 1),
+			FALSE,
+			TRUE, // raw literals
+			FALSE,
+			kind_for_define,
+			role_for_macro_undef,
+			kind_for_header,
+			role_for_header_system,
+			role_for_header_local
+		);
+
+	g_cxx.iChar = ' ';
+
+	boolean bRet = cxxParserParseBlock(FALSE);
+
+	cppTerminate ();
+
+	cxxTokenChainClear(g_cxx.pTokenChain);
+	if(g_cxx.pTemplateTokenChain)
+		cxxTokenChainClear(g_cxx.pTemplateTokenChain);
+	
+	if(!bRet && (passCount == 1))
+	{
+		CXX_DEBUG_PRINT("Processing failed: trying to rescan");
+		return RESCAN_FAILED;
+	}
+	
+	return RESCAN_NONE;
+}
+
+void cxxCppParserInitialize(const langType language)
+{
+	CXX_DEBUG_INIT();
+
+	CXX_DEBUG_PRINT("Parser initialize for language C++");
+	if(g_bFirstRun)
+		memset(&g_cxx,0,sizeof(CXXParserState));
+
+	g_cxx.eLanguage = language;
+	g_cxx.eCPPLanguage = language;
+	g_cxx.eCLanguage = -1;
+	cxxBuildKeywordHash(language,TRUE);
+}
+
+void cxxCParserInitialize(const langType language)
+{
+	CXX_DEBUG_INIT();
+
+	CXX_DEBUG_PRINT("Parser initialize for language C");
+	if(g_bFirstRun)
+		memset(&g_cxx,0,sizeof(CXXParserState));
+
+	g_cxx.eLanguage = language;
+	g_cxx.eCLanguage = language;
+	g_cxx.eCPPLanguage = -1;
+	cxxBuildKeywordHash(language,FALSE);
+}
+
+void cxxParserCleanup(void)
```

Comment:
Ow... you're absolutely right.
It is meant to be called in main, after all the files have been processed.
I've lost the call in one of the whole-repository reworks... added it back now.
Thnx! :)